### PR TITLE
Migrating PressableButton and PressableButtonHololens2 to use local positioning for movingButtonVisuals and iconText objects.

### LINKED ISSUE #7874: NaN exception when scaling pressable button parent to zero
## Describe the bug
Filing on behalf of Matt Rubinstein whos will be making a github account soon.

We're using PressableButtonHololens2 and noticed that we get a NaN exception on transforms attached to MovingButtonVisuals/MovingButtonIconText/CompressableButtonVisuals when any parent up-chain of them has a 0 scale.
 

image
 
We use a Hydration/Dehydration effect to scale up/down our prefabs on intro/outro which has them at 0 scale for a frame or two during the lifecycle. As a workaround, I modified our hydration effect to not go to exactly zero, but we still see this issue when values get too close to 0 scale.

## To reproduce

1. Set the scale of any parent of PressableButtonHoloLens 2 to zero
2. Observe NaN exceptions

## Release
MRTK 2.3

### PR DESCRIPTION
**Pull request by Zee2**
## Overview
An issue was reported (#7874) that when a PressableButton was scaled to zero, NaN values were propagated to a transform.position assign attempt. I analyzed this issue, and saw that the particular way that PressableButton calculates the visible button content offset uses a local-to-world scaling that is susceptible to NaN propagation when object scale is zero.

Upon further examination, it seems as though the current way that PressableButtons calculate the positioning of the `movingButtonVisuals` and `movingButtonIconText` children is slightly flawed; instead of using `localPosition` and relying on Unity to perform the appropriate transformations, the code performs the local-to-world scaling itself. While this behavior was correct for the majority of use cases, there was one weakness; when the button was initially scaled to zero, the local-to-world scaling suffered from NaN issues, as the world scaling would explode to infinity when the parent scale dropped to zero. Unity's internal implementation is apparently immune to these effects; assigning to localPosition, even when parent scale is zero, does not suffer from NaN explosions.

Several helpers were added to accomplish this migration from global position to local positioning; several "local" counterparts to the existing global initial position and push direction definitions were added, and naming was updated accordingly.

Unit test was added to verify that no NaN exceptions are thrown when a button is created with a parent with zero scale.

Therefore, this PR resolves #7874.

## Changes
- Adds private field `LocalSpacePressDirection`
- Adds private field `InitialLocalPosition`
- Renames `InitialPosition` to `InitialWorldPosition`
- Adds method `GetLocalPositionAlongPushDirection`, to complement the existing `GetWorldPositionAlongPushDirection`
- Changes the visual effect to use `localPosition` instead of `position`, using the new `GetLocalPositionAlongPushDirection` method.
- Changes `PressableButtonHololens2` to use localPosition for its unique effects, as well.
- Changes `PressableButtonHololens2` to properly check for divide-by-zero errors when setting the scale of the `compressableButtonVisuals`.
- Changes `PressableButton.GetDistanceAlongPushDirection` to _multiply_ by `WorldToLocalScale` instead of _dividing_ by `LocalToWorldScale`, which is more idiomatic and more closely aligns the code with the documented intent of the scaling factors.

## Verification
Unit test has been added, and run locally. I've also personally verified the behavior with the HoloLens 2 with both near and far interactions, with both the basic-style buttons and the HoloLens 2 Shell-style buttons.

### CODE REVIEW

**Comment by julenka** [comment]
@thalbern  I believe there was a reason why PressableButton used world space positions instead of local space positions, I believe either you or Daniel worked on this previously. Want to make sure you sign off on this because you spent quite a while early on with pressable button.

**Comment by Zee2** [comment]
I would agree that if we were allowing users to set the planes in global space before, we should maintain that feature; however, as far as I could tell in the code, there was only one code path for assigning the button visual displacement, which was:

press distance -> local vector -> project to global -> set global position -> Unity inverse transform -> rendered

Effectively, this change just lets Unity use its own local->global calculations, such that the flow is now

press distance -> local vector -> local position -> Unity transform system -> rendered.

By letting Unity handle the transforms, we don't need to rely on the divide-by-zero-susceptible global projection math. The Unity transform system is largely NaN safe when operating in zero-scaled coordinate systems/reference frames.

**Comment by thalbern** [comment]
> Thanks for having a look at this!
> 
> I haven't touched this code for a while, I'm currently testing your branch, but these are my thoughts:
> 
> In general i see the benefit of using the localPosition set of Unity but there's two things that bother me here:
> 
> 1. You've only modified one usage of the LocalToWorldScale but didn't adjust the other methods - now the PressableButton is left in an inconsistent state - where we previously had all calculations based on world space transforms it's now a mix of world and local space
> 2. The new GetLocalPositionAlongPushDirection is based under the assumption that all the distances passed in are in local space, will the button still work as expected once a user configures this for world space distances? (for example did you have a look at the inspector code that draws the planes, will that still behave correctly in both cases? will the button trigger correctly on setup of world space plane distances?)
> 
> We might not have enough tests to check the world space plane setup. I've only added conversion tests but I think the rest of our tests are based on local space distances, so they won't uncover if this is not working anymore.

Ok nevermind - I just realized that because you're only modifying the part that moves the visuals this is still working - inspector and other calulations are still based on the old system. Still wouldn't there be potential for those other spots to get into the same situation on a 0 scale?

image

So this fix solves one usage of that scale but not the others - We should either patch up all the code or add a null check to that scale value. 
What are your thoughts on this?

**Comment by thalbern** [comment]
did another test where i quickly added a script that would scale my gameobject to zero local scale and added it to the button. the method that scales is triggered on the onclick event when interacting with the button.
while doing that i'm getting these 

image

seems like this is still potentially resulting in a NaN

**Comment by Zee2** [comment]
@thalbern Changes have been added to resolve the DistanceSpaceMode in-editor toggle issue.

**Comment by Zee2** [comment]
I've added fixes based on feedback from @thalbern as well as changed `PressableButtonHololens2` to properly check for divide-by-zero when setting the compressible visuals scale. In addition, I changed `GetDistanceAlongPushDirection` to multiply by `WorldToLocalScale` instead of dividing by `LocalToWorldScale`, which is more idiomatic and more clearly aligns with the documented purpose of the scale factors.

**Comment by julenka** [comment]
@Zee2  I'm wondering, if @thalbern found errors after writing a script to scale button down to zero after click, is this a test case that the play mode test is not covering, that it should cover? Referring to this comment form @thalbern 

> did another test where i quickly added a script that would scale my gameobject to zero local scale and added it to the button. the method that scales is triggered on the onclick event when interacting with the button.

**Comment by Zee2** [comment]
> @Zee2 I'm wondering, if @thalbern found errors after writing a script to scale button down to zero after click, is this a test case that the play mode test is not covering, that it should cover? Referring to this comment form @thalbern
> 
> > did another test where i quickly added a script that would scale my gameobject to zero local scale and added it to the button. the method that scales is triggered on the onclick event when interacting with the button.

I re-tested this case; the play mode test should cover this. As long as the unit test/test runner fails the test on _**any**_ thrown exception, it should be caught. As far as I can tell, the `OnClick` is not a significant factor in this repro. Issue can be repro'd when scaling to zero before button init, and at any time during the button lifecycle when deleting the button objects.

File: Assets/MRTK/SDK/Features/UX/Scripts/PressableButtons/PressableButton.cs [hunk 1]
```diff
@@ -208,6 +209,24 @@ private Vector3 WorldSpacePressDirection
...
 
+        /// <summary>
+        /// The press direction of the button as defined by a NearInteractionTouchableSurface, in local space,
+        /// using Vector3.forward as an optional fallback when no NearInteractionTouchableSurface is defined.
+        /// </summary>
+        private Vector3 LocalSpacePressDirection
+        {
+            get
+            {
+                var nearInteractionTouchable = GetComponent<NearInteractionTouchableSurface>();
```
Comment (keveleigh) [inline Assets/MRTK/SDK/Features/UX/Scripts/PressableButtons/PressableButton.cs:220]: Since it looks like we call `GetComponent<NearInteractionTouchableSurface>();` both here and in `WorldSpacePressDirection`, every time they're called, it's probably worth caching the result and only trying to get the component again if the cached value turns null.

Comment (Zee2) [reply Assets/MRTK/SDK/Features/UX/Scripts/PressableButtons/PressableButton.cs:220]: @keveleigh this was something that I was going to change in a later PR. There are pretty egregious examples of non-cached references _all_ over the code, but I was going to put them in separately (discussed this with @thalbern)

Comment (keveleigh) [reply Assets/MRTK/SDK/Features/UX/Scripts/PressableButtons/PressableButton.cs:220]: @Zee2 Perfect, thanks!
